Fixes #5436 Randomize pretest questions

rohitkatlaa opened the pull request:
## Overview

<!--
READ ME FIRST: Please answer *both* questions below and check off every point from the essential checklist!
-->

1. This PR fixes #5436 .
2. The implementation details of the PR is in this [doc](https://docs.google.com/document/d/1IUhXcmEps7bJwFvPPBRxXYZ8tpl_zOoETKzQA_Q4vSw/edit?usp=sharing)

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

rohitkatlaa committed:
Merge pull request #17 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #18 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #19 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #20 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #21 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #22 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #23 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #24 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #25 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #26 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #27 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #28 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #29 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #30 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #31 from oppia/develop

Only expose thread data through fetchThreads (#8496)

rohitkatlaa committed:
Merge branch 'develop' of https://github.com/rohitkatlaa/oppia into develop

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

rohitkatlaa committed:
randomize pretest questions

rohitkatlaa committed:
failing backend tests

rohitkatlaa committed:
failing backend tests

rohitkatlaa committed:
failing backend tests

rohitkatlaa committed:
Merge branch 'randomize-pretest-questions' of https://github.com/rohitkatlaa/oppia into randomize-pretest-questions

rohitkatlaa committed:
backend tests

rohitkatlaa committed:
backend tests

rohitkatlaa committed:
spacing changes

rohitkatlaa commented:
I am sorry this PR took a long time to create because the backend tests were failing randomly at different places, So I had to run the backend tests several times to test if whether any tests are failing.

seanlip commented:
Hi @rohitkatlaa, backend tests should be deterministic. Could you please explain why they were failing "randomly"? If they indeed are failing randomly, this needs investigation.

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -316,25 +316,70 @@ def get_question_skill_links_by_skill_ids(
-                question_skill_count,
-                start_cursor=cursor
-            )
+            direction = start_cursor[-1]
```
This is really messy and has a lot of duplicated code. Please try and consolidate it more, and then I'll take another pass. Try to follow the DRY principle ("don't repeat yourself") -- this is very important for writing maintainable code.

rohitkatlaa replied:
I am not sure if it is possible to consolidate because for each case the query conditions are different.

seanlip replied:
There are two invocations of ">" and two invocations of "<=". Each of these are repeats and it should still be possible to consolidate. You might need to create additional functions that you can reuse.

rohitkatlaa replied:
I was able to reduce one such instance but I feel that the code not easy to understand. The other instance of  > has a difference in cursor which has to be in if condition so there has to be two instances.

vinitamurthi replied:
You can create a helper method to do this:
Something on the lines of: 
```
def get_question_order(direction, question_id, random_id):
    return question_id > random_id if direction == 1 else question_id <= random_id

def get_questions_list(skill_ids, random_id, question_skill_count, direction, cursor):
return cls.query(  cls.skill_id.IN(skill_ids)
                ).order(cls.question_id, cls.key).filter(
                    get_question_order(direction, cls.question_id, random_id)
                ).fetch_page(
                    question_skill_count,
                    start_cursor=cursor
                )
```

And then you can just call the helper function as explained below

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -316,25 +316,70 @@ def get_question_skill_links_by_skill_ids(
             next_cursor.urlsafe() if (next_cursor and more) else None
         )
+        if next_cursor_str:
+            next_cursor_str = (next_cursor_str +
```
Break after '('. See coding style guide (I mentioned this in a different PR review too).

rohitkatlaa replied:
Done.

seanlip commented:
@rohitkatlaa -- one general note, please follow the PR instructions [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change) when making a PR. In particular, when you want a reviewer to review something, make sure to add them to the assignees field. If you are not the next "actor" for a PR, remove yourself from the assignees field.

Please also note that your primary reviewer for this PR is @vinitamurthi.

rohitkatlaa commented:
@seanlip the tests for my code where failing randomly because I was using a random variable. So I had to run the entire backend tests several times to test it.

rohitkatlaa commented:
Is it fine if I just update the argument comments in only gae_models.py or should I update it where ever the cursor is used.

seanlip commented:
You need to find a way to make your tests deterministic. We cannot check in tests that depend on a random variable, that will break the CI for other developers.

If you are randomizing the order of a list, sort the actual and expected lists and check equality.

If you are randomizing the elements of a list, then do an "is one of" check in the expectation.

But the tests need to be written in such a way that they pass with 100% probability. Does that make sense?

> Is it fine if I just update the argument comments in only gae_models.py or should I update it where ever the cursor is used.

I am not sure what you are referring to here. Could you please elaborate?

rohitkatlaa commented:
The comment I mentioned earlier was about how I am appending the random id and the direction of fetch to the cursor.

rohitkatlaa commented:
Just to be clear the tests are not based on random variable, only the fetched values are based on it.
As you can see, I have already modified the tests to sort and assertIn. I just mentioned that the delay of making the PR was the tests.

rohitkatlaa committed:
reducing code

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -316,25 +316,58 @@ def get_question_skill_links_by_skill_ids(
-                question_skill_count,
-                start_cursor=cursor
-            )
+            direction = start_cursor[-1]
```
I don't understand the cursor logic and why you're doing indexing with -13, -1, etc. Could you please explain this? It is quite opaque.

rohitkatlaa replied:
I am attaching the random_id and direction of fetch to the cursor so that I can create exactly same query in the subsequent calls. The last char is the direction i.e 0 => '>' and 1 => '<=' and next last 12 char is the random id used to filter it. These details are required for subsequent queries otherwise the cursor can't be used.

vinitamurthi replied:
Even I am not completely getting the logic behind this, it might be worthwhile writing a doc explaining your logic. One more general question, I believe the number of pretest questions would be at the most 3-4, so why do we even need to fetch questions by page? Why can't we just pick up all questions and send it in one shot (like how we do it for practice sessions and review tests)? @aks681 @seanlip

vinitamurthi replied:
@rohitkatlaa Can you explain this logic before we go further in the review? It would be great to write up a small doc and give it for review to @aks681 @seanlip and I

rohitkatlaa replied:
@vinitamurthi @seanlip @aks681  I have updated the original [doc](https://docs.google.com/document/d/1IUhXcmEps7bJwFvPPBRxXYZ8tpl_zOoETKzQA_Q4vSw/edit?usp=sharing) design explaining the logic used. PTAL.

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -316,25 +316,58 @@ def get_question_skill_links_by_skill_ids(
-                start_cursor=cursor
-            )
+            direction = start_cursor[-1]
+            random_id = unicode(start_cursor[-13:-1]) #pylint: disable=unicode-builtin
```
You should not be doing unicode conversions (in general if you have to use a pylint pragma, something is probably going wrong).

Please explain the rationale (but I think you might actually need to find a different approach here).

rohitkatlaa replied:
I used unicode because the question id was in unicode and I thought that filtering required unicode conversion but it worrked without conversion. So i removed it.

seanlip commented:
@rohitkatlaa In the next review round, please assign the PR to @vinitamurthi. Thanks!

rohitkatlaa committed:
remove unicode

rohitkatlaa commented:
@vinitamurthi  PTAL.

vinitamurthi commented on core/storage/question/gae_models.py:
```diff
@@ -316,25 +316,58 @@ def get_question_skill_links_by_skill_ids(
+            direction = start_cursor[-1]
+            random_id = start_cursor[-13:-1]
+            cursor = datastore_query.Cursor(urlsafe=start_cursor[:-13])
+            if direction == '1':
+                questions_list, next_cursor, more = cls.query(
+                    cls.skill_id.IN(skill_ids)
+                ).order(cls.question_id, cls.key).filter(
+                    cls.question_id > random_id
+                ).fetch_page(
+                    question_skill_count,
+                    start_cursor=cursor
+                )
+            else:
+                questions_list, next_cursor, more = cls.query(
+                    cls.skill_id.IN(skill_ids)
+                ).order(cls.question_id, cls.key).filter(
+                    cls.question_id <= random_id
+                ).fetch_page(
+                    question_skill_count,
+                    start_cursor=cursor
+                )
```
You can replace this entire section with

```
questions_list, next_cursor, more = get_questions_list(skill_ids, random_id, question_skill_count, direction, cursor)
```

rohitkatlaa replied:
Done.

vinitamurthi commented on core/storage/question/gae_models.py:
```diff
@@ -316,25 +316,58 @@ def get_question_skill_links_by_skill_ids(
+                python_utils.UNICODE(
+                    utils.get_random_int(base_models.RAND_RANGE)),
+                base_models.ID_LENGTH)
+            questions_list, next_cursor, more = cls.query(
+                cls.skill_id.IN(skill_ids)
+            ).order(cls.question_id, cls.key).filter(
+                cls.question_id > random_id
+            ).fetch_page(question_skill_count)
```
You can replace this (here and below) with: 
```
questions_list, next_cursor, more = get_questions_list(skill_ids, random_id, question_skill_count, direction, cursor)
```

rohitkatlaa replied:
Done.

vinitamurthi commented on core/storage/question/gae_models_test.py:
```diff
@@ -294,16 +294,18 @@ def test_get_question_skill_links_by_skill_ids(self):
         self.assertEqual(len(question_skill_link_models), 2)
-        self.assertEqual(question_skill_link_models[0].skill_id, skill_id_2)
-        self.assertEqual(question_skill_link_models[1].skill_id, skill_id_1)
+        self.assertIn(
+            question_skill_link_models[0].skill_id, [skill_id_1, skill_id_2])
+        self.assertIn(
+            question_skill_link_models[1].skill_id, [skill_id_1, skill_id_2])
```
While all of these sorting and 'assertIn' functionality are making existing tests deterministic, we haven't actually tested that the shuffling is working as expected. Apart from these tests I believe we need another test where we fix a seed for the randomizer and then check the order, and then change the seed for the randomizer and ensure we have some other order and set of questions

rohitkatlaa replied:
I have added additional tests. PTAL.

rohitkatlaa committed:
adding additional tests and using helper functions

rohitkatlaa committed:
spelling

rohitkatlaa commented:
@vinitamurthi I have already updated the doc. PTAL
> @vinitamurthi @seanlip @aks681 I have updated the original [doc](https://docs.google.com/document/d/1IUhXcmEps7bJwFvPPBRxXYZ8tpl_zOoETKzQA_Q4vSw/edit?usp=sharing) design explaining the logic used. PTAL.

vinitamurthi commented:
> @vinitamurthi I have already updated the doc. PTAL
> 
> > @vinitamurthi @seanlip @aks681 I have updated the original [doc](https://docs.google.com/document/d/1IUhXcmEps7bJwFvPPBRxXYZ8tpl_zOoETKzQA_Q4vSw/edit?usp=sharing) design explaining the logic used. PTAL.

I still didn't understand why we need to specify indexes. Can we find a better way to do that? Also waiting on @seanlip and @aks681 for my earlier question:
> One more general question, I believe the number of pretest questions would be at the most 3-4, so why do we even need to fetch questions by page? Why can't we just pick up all questions and send it in one shot (like how we do it for practice sessions and review tests)?

seanlip commented:
> One more general question, I believe the number of pretest questions would be at the most 3-4, so why do we even need to fetch questions by page? Why can't we just pick up all questions and send it in one shot (like how we do it for practice sessions and review tests)?

Sorry, I missed this question. I don't believe we need to fetch pages of questions for the pretest, since as @vinitamurthi says, the number of questions is small (3-4).

But are we using this randomization functionality in other contexts like practice sessions too? If so, then maybe pagination is needed (if practice sessions become adaptive and the number of questions is unknown at the outset)? I'm not sure I fully understand the context of this PR -- perhaps @rohitkatlaa could clarify on why pagination is necessary.

rohitkatlaa commented:
@seanlip @vinitamurthi  Pagination is necessary only if we fetch more than 10-20 question which can be fetched in batches. If those many number of questions are going to fetched then it is preferable to send the questions in batches so that they don't take up too much bandwidth. But if the required number of questions is less then pagination is not necessary. So it all depends on whether we will fetch more questions or not. As far as pretest questions is concerned, at the most we will fetch only 3-4 questions.

rohitkatlaa commented:
I am sorry I clicked close and comment by mistake.

seanlip commented:
@rohitkatlaa Could you please clarify what your question for me is, please? Otherwise I don't know what to do when you assign me.

If your question for me is whether we need pagination for the pre-test, the answer is no based on what you said.

rohitkatlaa commented:
@seanlip So in that case should I remove the fetch_page and cursor?

seanlip commented:
@rohitkatlaa, I think you can make decisions on this based on the user requirements for what you are building.

rohitkatlaa commented:
@seanlip As a user I wouldn't want to attempt more than 3-4 pre test questions before a lesson. In that case as mentioned above there would be no benefit in fetching the questions in batches. So it is better to remove pagination and cursor. Am I correct?

seanlip commented:
Sure, seems reasonable to me.

rohitkatlaa commented:
Eventhough we reached a conclusion that we will remove the cursor and pagination, I have been thinking that since we already have pagination we can leave it like that since it can be used if we require in the future where we will fetch many questions
Am I wrong in thinking in this way?

seanlip commented:
@rohitkatlaa, you should be checking with @vinitamurthi, not me.

@vinitamurthi, will we ever need to fetch many questions with this functionality?

rohitkatlaa commented:
@seanlip @vinitamurthi I had a talk with @aks681 about this and he reminded me that the pagination will be useful when displaying the set of questions in the skill editor. There the questions is fetched in batches of 10. Here the cursor and pagination is used. So I think we should not remove the pagination and cursor.

seanlip commented:
@rohitkatlaa -- this is a decision for @aks681 and @vinitamurthi. Please consult them.

rohitkatlaa commented:
@seanlip, as per @aks681 advice I have already started a mail thread for discussing this topic. But a decision is yet to be concluded. Will update once a decision is reached.

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into randomize-pretest-questions

vinitamurthi commented:
@rohitkatlaa You can change your implementation as discussed in the email

nithusha21 commented:
Hi @rohitkatlaa could you please merge with develop as there were some new checks introduced, and a change to the development workflow which will help make sure develop will not break after merging this PR.

rohitkatlaa committed:
randomize the pretest questions and practice session questions

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into randomize-pretest-questions

rohitkatlaa committed:
linting

vinitamurthi commented on core/storage/question/gae_models.py:
```diff
@@ -365,51 +362,67 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
             # number requested, even after deduplication.
-            new_question_skill_link_models = (
-                equal_questions_query.fetch(question_count_per_skill * 2))
+            new_question_skill_link_models = equal_questions_query.fetch()
```
Why did you remove the extra fetch of questions? We still have to ensure that we remove duplicates right? We should first remove duplicates and then do the randomization

rohitkatlaa replied:
By removing the count I am fetching all the questions that match the difficulty(according to your suggestion) and randomly picking out of them. So essentially I am fetching more than question_count_per_skill * 2.

vinitamurthi replied:
Oh sorry I misunderstood, but I still don't understand why we are fetching all instead of double. @sophiewu6 had made the fetch to double instead of all because the performance drops significantly if we iterate over all records. @sophiewu6 do you have any thoughts?

sophiewu6 replied:
Fetching all questions can improve the randomness of selecting questions. However, there could be a large number of questions for one skill. In that case, fetching all questions could significantly lower the performance. @rohitkatlaa Have you evaluated its performance before and after your change?

seanlip replied:
I haven't been following this PR, but I saw @sophiewu6's comment and want to highlight that you absolutely cannot fetch *all* questions. There needs to be a reasonable upper bound because there may be thousands of questions in the datastore.

rohitkatlaa replied:
@sophiewu6  I haven't evaluated it's performance but what can be done is to fetch the questions with question id greater than a particular random id.(this was the method I used earlier)

rohitkatlaa replied:
@seanlip @sophiewu6  Just to be clear I am only fetching all the questions within a particular skill_id with particular difficulty, and I think that is going to be limited. And the earlier method of using a random_id to fetch questions with question_id greater than the random_id cannot be used in this [function](https://github.com/oppia/oppia/blob/8d8689bc45a75d5f89025ff216a2eccea1c9c2e9/core/storage/question/gae_models.py#L327) because google [datastore](https://cloud.google.com/appengine/docs/standard/python/datastore/query-restrictions#inequality_filters_are_limited_to_at_most_one_property) allows only one inequality filter at a time.

seanlip replied:
There may easily be > 100 questions for a particular skill ID with a particular difficulty. So this method still doesn't work.

Perhaps you can use a random offset? I.e. generate the offset number for the query randomly?

rohitkatlaa replied:
As I have mentioned in the earlier comment it is not possible to use two inequality filters(difficulty and random offset) at the same time.

sophiewu6 replied:
I also noticed that problem last year. I have made a google doc for possible solutions. Maybe you can take a look? https://docs.google.com/document/d/1nghfz229A65qQGen57kwQrHlQ4qQ6nFMD0_bqdVrj0A/edit?usp=sharing

rohitkatlaa replied:
I had a look at the doc, both the solutions requires two inequality filters(difficulty and random id) for this [function](https://github.com/oppia/oppia/blob/8d8689bc45a75d5f89025ff216a2eccea1c9c2e9/core/storage/question/gae_models.py#L327).

seanlip replied:
Why does the random offset need to be an inequality filter? I was talking about the offset arg in https://cloud.google.com/appengine/docs/standard/python/datastore/gqlqueryclass#GqlQuery_fetch

rohitkatlaa replied:
I am sorry I didn't understand what offset meant. I think the offset can be used to fetch the random results.

seanlip replied:
I don't understand your response, sorry. Are you still stuck or does the clarification help?

rohitkatlaa replied:
I am sorry earlier I didn't understand the meaning of offset. I think the offset can be used to fetch random results.

seanlip replied:
Hi @rohitkatlaa, please do not delete earlier responses. If you do, then it becomes hard to follow the conversation.

Could you please clarify whether you are still stuck, or whether the previous clarification helps?

rohitkatlaa replied:
I understood the concept and I pushed the changes. PTAL

vinitamurthi commented on core/storage/question/gae_models.py:
```diff
@@ -365,51 +362,67 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
                 # requested difficulty and sort them by decreasing difficulty.
                 easier_questions_query = query.filter(
                     cls.skill_difficulty < difficulty_requested)
                 easier_questions_query = easier_questions_query.order(
```
I guess we dont need this ordering anymore right

rohitkatlaa replied:
Done.

vinitamurthi commented on core/storage/question/gae_models.py:
```diff
@@ -365,51 +362,67 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
                         cls.skill_difficulty)
                     harder_question_skill_link_models = (
-                        harder_questions_query.fetch(question_count_per_skill))
+                        harder_questions_query.fetch())
```
Can you explain why you are removing question counts per skill?

vinitamurthi commented on core/storage/question/gae_models_test.py:
```diff
@@ -388,6 +388,56 @@ def test_get_question_skill_links_based_on_difficulty(self):
+        )
+        self.assertEqual(len(question_skill_links), 2)
+        self.assertTrue(questionskilllink_model2 in question_skill_links)
+        self.assertTrue(questionskilllink_model1 in question_skill_links or
```
By testing this way you aren't really checking if the randomization is working as expected. Here's what you can do - set a random seed, call the function, verify the output. Change the random seed, call the function, verify that the output is now different

rohitkatlaa replied:
Done.

vinitamurthi commented on core/storage/question/gae_models_test.py:
```diff
@@ -388,6 +388,56 @@ def test_get_question_skill_links_based_on_difficulty(self):
+        self.assertTrue(questionskilllink_model1 in question_skill_links)
+        self.assertTrue(questionskilllink_model2 in question_skill_links)
+        self.assertTrue(questionskilllink_model3 in question_skill_links)
+        self.assertTrue(questionskilllink_model4 in question_skill_links or
```
Same comment about testing random functions applies here

rohitkatlaa replied:
Done.

vinitamurthi commented on core/storage/question/gae_models_test.py:
```diff
@@ -492,6 +542,38 @@ def test_get_question_skill_links_equidistributed_by_skill(
+            )
+        )
+        self.assertEqual(len(question_skill_links), 2)
+        self.assertTrue(questionskilllink_model1 in question_skill_links or
+                        questionskilllink_model2 in question_skill_links)
+        self.assertTrue(questionskilllink_model3 in question_skill_links or
+                        questionskilllink_model4 in question_skill_links)
```
See above comments about testing random functions and change the test

rohitkatlaa replied:
Done.

vinitamurthi reviewed (changes requested):
The overall direction of this is much better! Few implementation comments but design wise this looks good. Thanks @rohitkatlaa !

rohitkatlaa committed:
testing random fetches

vinitamurthi commented on core/storage/question/gae_models_test.py:
```diff
@@ -388,6 +389,49 @@ def test_get_question_skill_links_based_on_difficulty(self):
+        )
+        self.assertEqual(len(question_skill_links_1), 3)
+        self.assertEqual(len(question_skill_links_2), 3)
+        self.assertNotEqual(question_skill_links_1, question_skill_links_2)
```
While this is ensuring the two aren't equal, it would be good to also verify the contents of question_skill_links_1 and question_skill_links_2

rohitkatlaa replied:
The contents of question_skill_links_1 and question_skill_links_2 might be partially same so a one to one check might not be possible so I am not sure what you mean by verify the contents.

vinitamurthi replied:
So question_skill_links_1 and question_skill_links_2 would contain 3 out of the 5 models you have created above right?
What I am trying to say is do an assertItemsEqual(question_skill_links_1, [all the models that question_skill_links_1 should contain])
Same thing for question_skill_links_2

rohitkatlaa replied:
Done.

vinitamurthi commented on core/storage/question/gae_models_test.py:
```diff
@@ -492,6 +536,49 @@ def test_get_question_skill_links_equidistributed_by_skill(
+        )
+        self.assertEqual(len(question_skill_links_1), 3)
+        self.assertEqual(len(question_skill_links_2), 3)
+        self.assertNotEqual(question_skill_links_1, question_skill_links_2)
```
Same comment applies here. Verify the contents of the response as well, not just that the two aren't the same

rohitkatlaa replied:
Done.

rohitkatlaa committed:
adding additional verification

rohitkatlaa commented:
@sophiewu6 Ping.

sophiewu6 reviewed (commented):
Hi @rohitkatlaa Thanks for your PR! I took a brief pass. It seems to me that the basic logic of this PR is to fetch all questions and get random samples from all questions fetched. However, as @vinitamurthi said, this method would significantly lower its performance. Please correct me if I misunderstood. Thank you!

rohitkatlaa committed:
adding offset to randomisation

rohitkatlaa commented:
I changed the utils.get_random_int so that the seed can be set and it can be tested. Using a seed does not affect system random.

sophiewu6 commented on core/storage/question/gae_models.py:
```diff
@@ -354,15 +354,26 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
+
         for skill_id in skill_ids:
-            query = cls.query(cls.skill_id == skill_id)
+            query = cls.gql("WHERE skill_id = '" + skill_id + "'") # pylint: disable=invalid-string-quote
```
Sorry, what's the point of changing this?

rohitkatlaa replied:
I am sorry I didn't realize that query also had an offset. Done.

sophiewu6 commented on core/storage/question/gae_models.py:
```diff
@@ -354,15 +354,26 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
 
             equal_questions_query = query.filter(
-                cls.skill_difficulty == difficulty_requested)
+                ndb.FloatProperty('skill_difficulty') == difficulty_requested)
```
ditto

rohitkatlaa replied:
Done.

sophiewu6 commented on core/storage/question/gae_models.py:
```diff
@@ -354,15 +354,26 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
 
+        def get_offset(query):
+            """Helper function to get the offset."""
+            question_count = query.count()
```
From the documentation:
> count (read_policy=STRONG_CONSISTENCY, deadline=60, offset=0, limit=1000, start_cursor=None, end_cursor=None)
> Returns the number of results matching the query. This is faster by a constant factor than actually retrieving all of the results, but the running time still scales linearly with the sum of offset + limit. Unless the result count is expected to be small, it is best to specify a limit argument; otherwise the method will continue until it finishes counting or times out.

It seems like the running time of this method still scales linearly, which could be concerning. It is true that it's faster than retrieving all results. @seanlip Is this way good enough?

seanlip replied:
I think it's probably OK. Let's go with it for now. But make sure there's a limit. Also do some manual testing to see what the performance times are and how they scale with different numbers of items in the datastore.

rohitkatlaa replied:
Is there a particular way I can test, one way is to create the questions and fetch them in practice session but I am not sure how to time the event.

seanlip replied:
Just noting, for posterity, that the thread is continued [here](https://github.com/oppia/oppia/pull/9061#issuecomment-629765809). (Sorry for breaking it off into the main thread -- github doesn't seem to have a way to "reply to thread" by email.)

sophiewu6 reviewed (commented):
Thanks, @rohitkatlaa. I took a brief pass again.

rohitkatlaa committed:
requested changes

seanlip commented:
You can add some logging functions in Python to log the current time
before/after and take the difference.

>

rohitkatlaa commented:
Randomized pretest questions testing
- 1000 similar questions with same skill id with same difficulty and fetch 100 questions --- 0.54768s
- 10000 similar questions with same skill id with same difficulty and fetch 1000 questions --- 6.309633s
- 1000 similar questions with 5 different skill id with same difficulty and fetch 100 questions --- 0.485407s
- 10000 similar questions with 5 different skill id with same difficulty and fetch 1000 questions --- 5.5707s
- 1000 similar questions with 5 different skill id with different difficulty and fetch 100 questions --- 0.55987s
- 10000 similar questions with 5 different skill id with different difficulty and fetch 1000 questions --- 6.01591s

The original code with 10000 question and fetching 1000 questions took approximately 5.339581s and the one with 1000 questions and fetching 100 question took approximately 0.468296s

In conclusion there is a slight increase in the time taken but I think it is not substantial to cause much of a difference.

seanlip commented:
Thanks for the analysis @rohitkatlaa! This is very helpful.

One further question: in each of the scenarios above, what is the time needed for fetching (a) 5 questions, (b) 20 questions? These numbers would be more typical for pre-tests and practice sessions.

seanlip commented:
Also, for each of the question fetches, could you please specify how many skills they are fetching questions for? (1 or many?) How does the time scale with "number of skills requested"?

rohitkatlaa commented:
Randomized pretest questions testing

- 10000 similar questions with 10 different skill id with different difficulty and fetch 5 questions  with same skill id  --- 0.686908s
- 10000 similar questions with 5 different skill id with different difficulty and fetch 20 questions with same skill id --- 0.714442s
- 10000 similar questions with 10 different skill id with different difficulty and fetch 5 questions with three different skill id --- 2.125972s
- 10000 similar questions with 10 different skill id with different difficulty and fetch 20 questions with three different skill id --- 2.35117s
- 10000 similar questions with 10 different skill id with different difficulty and fetch 100 questions with three different skill id --- 3.134768s
- 10000 similar questions with 10 different skill id with different difficulty and fetch 1000 questions with three different skill id --- 4.88561s

Another observation is that the time taken by tests seems to reduce in the subsequent fetches i.e the first fetch takes a little more time than the second fetch.

seanlip commented:
Thanks, this is great. The time reduction in the subsequent fetches might be due to some kind of caching at the datastore layer, though I didn't think this would happen for queries.

I think these times are reasonable. Could you please do one more check: what is the timing for 10000 similar questions with 500 different skill ids with different difficulty, where we fetch (a) 5 (b) 20 (c) 1000 questions with three different skill ids? I am expecting about 20 questions per skill ID and I think 500 skills is reasonable; I'm hoping that the increased number of skill IDs doesn't cause much of an increase in the time taken.

rohitkatlaa commented:
**Trial-1**
- 10000 similar questions with 500 different skill id with different difficulty and fetch 5 questions with three different skill id --- 4.067628s
- 10000 similar questions with 500 different skill id with different difficulty and fetch 20 questions with three different skill id --- 5.0325179s
- 10000 similar questions with 500 different skill id with different difficulty and fetch 100 questions with three different skill id --- 8.425033s

**Trial-2**
- 10000 similar questions with 500 different skill id with different difficulty and fetch 5 questions with three different skill id --- 2.9777833s
- 10000 similar questions with 500 different skill id with different difficulty and fetch 20 questions with three different skill id --- 4.3930411s
- 10000 similar questions with 500 different skill id with different difficulty and fetch 100 questions with three different skill id --- 7.6122400s

**Trial-3**
- 10000 similar questions with 500 different skill id with different difficulty and fetch 5 questions with three different skill id --- 3.481895s
- 10000 similar questions with 500 different skill id with different difficulty and fetch 20 questions with three different skill id --- 4.3087902s
- 10000 similar questions with 500 different skill id with different difficulty and fetch 100 questions with three different skill id --- 7.3470630s

There is indeed a increase in time taken to fetch the questions.

seanlip commented:
The timing does worry me a little. I think "10000 similar questions with 500 different skill id with different difficulty and fetch 20 questions with three different skill id" is reasonably standard for practice sessions.

We may need to do something more complex later, but for now I'm fine with this approach. @vinitamurthi what do you think -- is there anything better? Also, @rohitkatlaa, are there other approaches that you've considered -- if so, it might be worth trying to time them? (The time estimates really help, thanks!)

rohitkatlaa commented:
@seanlip I don't know any other approaches. The solutions in the doc provided by @sophiewu6 also are not feasible because datastore does not support multiple inequality filters.

seanlip commented:
OK. Let's go with this approach for now, then. @rohitkatlaa -- in the part which implements the controversial query, could you please add a long comment describing clearly what we've tried, why we selected this one, and pointing readers to this thread? I suspect this will end up needing to be revisited in future, so let's leave a breadcrumb trail in the code so that developers can understand why the decision was made. Thanks!

vinitamurthi commented:
@seanlip @aks681 Regarding future improvements, I recall there being a discussion a while ago about creating a performant version of the QuestionSkillLink store. I don't know if there was any work done on that, but maybe something to consider.

vinitamurthi commented on core/storage/question/gae_models.py:
```diff
@@ -357,59 +354,88 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
+                    new_question_skill_link_models, question_count_per_skill)
+            else:
                 # Fetch QuestionSkillLinkModels with difficulty smaller than
                 # requested difficulty and sort them by decreasing difficulty.
```
I think we need to change these comments considering we are not ordering them by skill difficulty anymore

rohitkatlaa replied:
Done.

seanlip commented:
> @seanlip @aks681 Regarding future improvements, I recall there being a discussion a while ago about creating a performant version of the QuestionSkillLink store. I don't know if there was any work done on that, but maybe something to consider.

Hi @vinitamurthi -- do we have any docs/etc. about this? I've actually completely forgotten that discussion and I'm not sure what the outcome of it was. Maybe worth starting an email thread?

rohitkatlaa committed:
adding a detailed comment

seanlip commented on core/controllers/reader.py:
```diff
@@ -265,24 +264,22 @@ class PretestHandler(base.BaseHandler):
                 feconf.NUM_PRETEST_QUESTIONS,
                 story.get_prerequisite_skill_ids_for_exp_id(exploration_id),
-                start_cursor)
+                False)
```
Would you mind changing the name of the third arg in this function to require_medium_difficulty? That seems to be what is actually implemented if I'm reading it right...

(And in that case I think it's fine to have this be True -- it make sense for all pre-test questions to be medium difficulty.)

rohitkatlaa replied:
Done.

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -357,59 +354,97 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
+        # Other alternative methods were:
+        # 1) Using a random id in question id filter
+        # 2) Adding an additional column that can be filtered upon.
+        # But these methods are not viable because google datastore limit
```
... limits each query to have at most one inequality filter. ...

(Great comment, btw!)

rohitkatlaa replied:
Done.

seanlip commented on utils.py:
```diff
@@ -259,6 +258,24 @@ def get_random_choice(alist):
     return alist[index]
 
 
+def get_random_sample(alist, sample_length):
```
I don't see why you need to break this out into utils.py. Why not just do random.sample() directly in the code?

rohitkatlaa replied:
I did because nowhere in the code they used it directly. But now I have removed it and used it directly.

seanlip commented on utils.py:
```diff
@@ -240,8 +240,7 @@ def get_random_int(upper_bound):
     """
     assert upper_bound >= 0 and isinstance(upper_bound, int)
 
-    generator = random.SystemRandom()
```
Why remove this?

rohitkatlaa replied:
If a generator is used I won't be able to set a seed so in order to set the seed and test I removed it.

seanlip replied:
This is no longer an issue, right? Or if it is, you should mock it.

As a general rule, **please do not change code in core utility functions** just to make the tests on your PR pass. You risk breaking other parts of the application.

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -450,21 +483,35 @@ def get_question_skill_links_equidistributed_by_skill(
                 if model.question_id in existing_question_ids:
                     new_question_skill_link_models.remove(model)
+            if len(new_question_skill_link_models) > question_count_per_skill:
+                new_question_skill_link_models = utils.get_random_sample(
```
Do not redefine variables. Choose a different name instead.

rohitkatlaa replied:
Done.

seanlip reviewed (changes requested):
Took a pass, it's looking pretty good IMO :) Thanks @rohitkatlaa!

rohitkatlaa committed:
requested changes

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into randomize-pretest-questions

marianadasilvadev reviewed (approved):
LGTM as codeowner perspective.

seanlip reviewed (changes requested):
@rohitkatlaa Did you test your changes manually?

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -389,7 +390,7 @@ def get_offset(query):
 
             if len(new_question_skill_link_models) >= question_count_per_skill:
-                new_question_skill_link_models = utils.get_random_sample(
+                new_question_skill_link_models = utils.random.sample(
```
This looks completely wrong: utils doesn't have a member named random. But could you please investigate why the backend tests are not catching it, and appear to be passing despite this?

rohitkatlaa replied:
I have tested this and it indeed works. What this does is import the random that is imported by the utils file. This is how I set the seed in the gae_models_test file. Since you asked me to use random.sample I thought it would be better if I use the one from utils file. I will change it if this method is wrong.

seanlip replied:
Oh, I see, don't do that please. Instead, in the test file, swap out the call to random.sample with your own implementation.

seanlip replied:
(Also, thanks for explaining!)

rohitkatlaa replied:
Just to be clear you want me to implement random.sample (like I did previously in the utils file). Or you want me to directly import random in gae_models and use random.sample?

seanlip replied:
Neither. In the tests, use self.swap() to swap random.sample() with a mock method that does something deterministic.

rohitkatlaa replied:
If we use random.sample directly I don't think it is possible to swap the method with the mock method.
i.e self.swap(random,'sample', mock_random_sample) doesn't work

seanlip replied:
It should work. Why doesn't it work?

rohitkatlaa replied:
I am not sure maybe because it is a python inbuilt library and not one which we implemented. But instead of using swap we can set the seed for random right, It should work fine.

seanlip replied:
@rohitkatlaa -- when you say "X doesn't work", please always, at the very least, provide the code you tried, and the stack trace, and what you did to try and figure out where the error in the stack trace comes from. Otherwise, I won't be able to help you, because I don't know what you have actually tried to do. (I can't tell at the moment whether the "doesn't work" is due to an actual issue or an error in implementation.)

rohitkatlaa replied:
I tied the following
```
def mock_random_sample(alist, num):
   if num >= len(alist):
        return alist
        alist.sort(key= question_id)
    return [1,2,3]

self.swap(random, 'sample', mock_random_sample)
```
Now the random.sample function is supposed to return [1,2,3] but instead it executed in the default way. I have also checked that the random.sample is being reached in the code. Thus I came to a conclusion that maybe inbuilt functions can't be swapped. A possible workaround for this can be to implement our sample function in the utils and swap it in the test.

vinitamurthi commented:
> > @seanlip @aks681 Regarding future improvements, I recall there being a discussion a while ago about creating a performant version of the QuestionSkillLink store. I don't know if there was any work done on that, but maybe something to consider.
> 
> Hi @vinitamurthi -- do we have any docs/etc. about this? I've actually completely forgotten that discussion and I'm not sure what the outcome of it was. Maybe worth starting an email thread?

I'll start one!

seanlip commented:
Still not sure about your code snippet. When you try mocking another
function (not a stdlib one), does it work as expected? Are you following
the pattern used in other tests for implementing swaps?

>
(When answering the above questions please provide enough detail about what
you tried so I can replicate it.)

rohitkatlaa committed:
adding tests

rohitkatlaa commented:
I am sorry I couldn't explain properly and implement it correctly. I failed to add a with statement.

seanlip commented on core/storage/question/gae_models.py:
```diff
@@ -357,59 +355,98 @@ def get_question_skill_links_based_on_difficulty_equidistributed_by_skill(
+        # each query to have at most one inequality filter. So we can't filter
+        # on both question_id and difficulty. Just noting, for posterity, that
+        # the thread is continued
+        # https://github.com/oppia/oppia/pull/9061#issuecomment-629765809 .
```
This last sentence should be changed to "Please see `https://github.com/oppia/oppia/pull/9061` for more details."

rohitkatlaa replied:
Done.

seanlip commented on core/storage/question/gae_models_test.py:
```diff
@@ -388,6 +390,81 @@ def test_get_question_skill_links_based_on_difficulty(self):
+            return alist[:num]
+
+        sample_swap = self.swap(random, 'sample', mock_random_sample)
+        utils.random.seed(1)
```
Please drop all calls to utils.random.

rohitkatlaa replied:
Done.

seanlip commented:
@vinitamurthi @aks681 I think this is ready for codeowner review. Thanks!

rohitkatlaa committed:
requested changes

vinitamurthi reviewed (approved):
This looks great now, thanks @rohitkatlaa !

seanlip reviewed (approved):
Looks good. Thanks @rohitkatlaa! 

@aks681 PTAL as codeowner.

aks681 reviewed (approved):
Lgtm as codeowner
